pkg/datapath/tables/node_address_test.go

```diff
@@ -0,0 +1,453 @@
+		},
+
+		wantNodePort: []net.IP{
+			net.ParseIP("10.0.1.1"),
+		},
+	},
+
+	{
+		name: "multiple",
+		addrs: []tables.DeviceAddress{
+			{
+				Addr:  netip.MustParseAddr("10.0.0.1"),
+				Scope: unix.RT_SCOPE_UNIVERSE,
+			},
+			{
+				Addr:      netip.MustParseAddr("10.0.0.2"),
+				Scope:     unix.RT_SCOPE_UNIVERSE,
+				Secondary: true,
+			},
+		},
+
+		wantLocal: []net.IP{
+			ciliumHostIP,
+			net.ParseIP("10.0.0.1"),
+			net.ParseIP("10.0.0.2"),
+		},
+
+		wantNodePort: []net.IP{
+			net.ParseIP("10.0.0.1"),
+		},
+	},
+	{
+		name: "ipv6 multiple",
+		addrs: []tables.DeviceAddress{
+			{
+				Addr:  netip.MustParseAddr("2600:beef::2"),
+				Scope: unix.RT_SCOPE_SITE,
+			},
+			{
+				Addr:  netip.MustParseAddr("2001:db8::1"),
+				Scope: unix.RT_SCOPE_UNIVERSE,
+			},
+		},
+
+		wantLocal: []net.IP{
+			ciliumHostIP,
+			net.ParseIP("2001:db8::1"),
+			net.ParseIP("2600:beef::2"),
+		},
+
+		wantNodePort: []net.IP{
+			net.ParseIP("2001:db8::1"),
+		},
+	},
+}
+
+func TestNodeAddress(t *testing.T) {
+	for _, tt := range nodeAddressTests {
+		t.Run(tt.name, func(t *testing.T) {
+			var (
+				db        *statedb.DB
+				devices   statedb.RWTable[*tables.Device]
+				nodeAddrs statedb.Table[tables.NodeAddress]
+			)
+			h := hive.New(
+				job.Cell,
+				statedb.Cell,
+				tables.NodeAddressCell,
+				tables.DeviceTableCell,
+				cell.Provide(func(t statedb.RWTable[*tables.Device]) statedb.Table[*tables.Device] { return t }),
+				cell.Invoke(func(db_ *statedb.DB, d statedb.RWTable[*tables.Device], na statedb.Table[tables.NodeAddress]) {
+					db = db_
+					devices = d
+					nodeAddrs = na
+				}),
+
+				// option.DaemonConfig needed for AddressMaxScope. This flag will move into NodeAddressConfig
+				// in a follow-up PR.
+				cell.Provide(func() *option.DaemonConfig {
+					return &option.DaemonConfig{
+						AddressScopeMax: defaults.AddressScopeMax,
+					}
+				}),
+			)
+
+			require.NoError(t, h.Start(context.TODO()), "Start")
+
+			txn := db.WriteTxn(devices)
+			_, watch := nodeAddrs.All(txn)
+
+			devices.Insert(txn, &tables.Device{
+				Index: 1,
+				Name:  "cilium_host",
+				Flags: net.FlagUp,
+				Addrs: []tables.DeviceAddress{
+					{Addr: ip.MustAddrFromIP(ciliumHostIP), Scope: unix.RT_SCOPE_UNIVERSE},
+				},
+				Selected: false,
+			})
+
+			shuffleSlice(tt.addrs) // For extra bit of randomness
+			devices.Insert(txn,
+				&tables.Device{
+					Index:    2,
+					Name:     "test",
+					Selected: true,
+					Flags:    net.FlagUp,
+					Addrs:    tt.addrs,
+				})
+
+			txn.Commit()
+			<-watch // wait for propagation
+
+			iter, _ := nodeAddrs.All(db.ReadTxn())
+			local := []string{}
+			nodePort := []string{}
+			for addr, _, ok := iter.Next(); ok; addr, _, ok = iter.Next() {
+				local = append(local, addr.Addr.String())
+				if addr.NodePort {
+					nodePort = append(nodePort, addr.Addr.String())
+				}
+			}
+			require.ElementsMatch(t, local, ipStrings(tt.wantLocal), "LocalAddresses do not match")
```

Comment:
nit: these should be asserts, it's what the test is testing, right?

Reply:
True, here it makes sense to assert and make sure to also stop!